src/shared/components/upload/Upload.test.tsx

```diff
@@ -15,130 +15,176 @@
+      expect(input?.hasAttribute('multiple')).toBe(false);
     });
 
-    it('disables multiple file selection if allowMultiple is set to false', () => {
-      const container = renderUpload({ allowMultiple: false }).container;
-      expect(
-        container.querySelector('input')?.getAttribute('multiple')
-      ).toBeFalsy();
+    it('allows upload of multiple files', () => {
+      const { container } = render(
+        <Upload onUpload={noop} allowMultiple={true} />
+      );
+      const input = container.querySelector('input');
+      expect(input?.hasAttribute('multiple')).toBe(true);
     });
+  });
+
+  describe('file uploading via file input', () => {
+    const onUpload = jest.fn();
+
+    it('uploads a single file', () => {
+      const { container } = render(<Upload onUpload={onUpload} />);
 
-    it('calls the windowService.getFileReader for every file selected', () => {
-      const container = renderUpload().container;
       fireEvent.change(container.querySelector('input') as HTMLElement, {
-        ...event,
-        target: { files: files }
+        target: { files: [file1] }
       });
-      expect(windowService.getFileReader).toHaveBeenCalledTimes(files.length);
+
+      const uploadedFile = onUpload.mock.calls[0][0] as File;
+
+      expect(uploadedFile.name).toBe('file1.txt');
     });
 
-    it('calls the readAsText multiple times based on the number of files', () => {
-      const container = renderUpload().container;
+    it('uploads multiple files', () => {
+      const { container } = render(
+        <Upload onUpload={onUpload} allowMultiple={true} />
+      );
+
       fireEvent.change(container.querySelector('input') as HTMLElement, {
-        ...event,
-        target: { files: files }
+        target: { files: [file1, file2] }
       });
-      expect(mockFileReader.readAsText).toHaveBeenCalledTimes(files.length);
+
+      const [uploadedFile1, uploadedFile2] = onUpload.mock
+        .calls[0][0] as File[];
+
+      expect(uploadedFile1.name).toBe('file1.txt');
+      expect(uploadedFile2.name).toBe('file2.txt');
     });
-  });
 
-  describe('using drag and drop', () => {
-    it('changes its appearance when the user drags a file over', () => {
-      const item = {};
-      const dataTransfer = {
-        items: [item]
-      };
+    // for tests of all possible file transformations, see uploadHelpers.test.ts
+    it('can read text contents of a file', async () => {
+      // intentionally redefining onUpload to prove
+      // that the type of the callback function matches component's expectations
+      let textFromFile = '';
+      const onUpload = (result: FileTransformedToString) =>
+        (textFromFile = result.content as string);
+
+      const { container } = render(
+        <Upload onUpload={onUpload} transformTo="text" />
+      );
+
+      await act(async () => {
+        fireEvent.change(container.querySelector('input') as HTMLElement, {
+          target: { files: [file1, file2] }
```

Comment:
Was it intentional to upload two files here?

Reply:
Kinda. I didn't have a strong opinion one way or the other. On the one hand, realistically, the event won't contain more than one file in the array, because the input element won't allow the user to select more than one file. On the other hand, we are proving to ourselves that even if, miraculously, the event did contain more than one file, only the first file would be processed. I can only keep one file in the array if you prefer 🤷‍♂️

Reply:
I was wondering what if the user drops multiple files but we only handle the first file, ignoring the rest?

Reply:
That's the current behaviour of the component: it will accept all files if the "allowMultiple" parameter is set to true; but will only process the first file if the allowMultiple parameter is set to false. I don't know what the correct behaviour should otherwise be if the allowMultiple is false, but the user drops multiple files.